caneda-help/docs/man/canedaattenuator.1
.TH CanedaAttenuator "1" "July 2006" "Debian/GNU Linux" "User Commands"
.SH NAME
CanedaAttenuator \- An attenuator synthesis application.
.SH SYNOPSIS
.B canedaattenuator
[\fIOPTION\fR] ...
.SH DESCRIPTION

\fBCaneda\fR is an integrated circuit simulator which means you will be
able to setup a circuit with a graphical user interface (GUI) and
simulate the large-signal, small-signal and noise behaviour of the
circuit.  After that simulation has finished you will be able to
present the simulation results on a presentation page or window.

The software aims to support all kinds of circuit simulation types,
e.g. DC, AC, S-parameter, harmonic balance analysis, noise analysis,
etc.

\fBCanedaAttenuator\fR is the attenuator synthesis tool used by Caneda.
By use of an input dialog the user can create an attenuator which is
then copied into the system-wide clipboard.  In \fBCaneda\fR the user
opens an empty schematic and presses CTRL-V (paste from
clipboard). The attenuator schematic is now inserted and can be
simulated.

Available attenuator topologies types are: Tee, Pi and Bridged-Tee.

.SH AVAILABILITY
The latest version of Caneda can always be obtained from
\fBwww.caneda.tuxfamily.org\fR
.SH "REPORTING BUGS"
Known bugs are documented within the BUGS file.  Report bugs to
<caneda-bugs@lists.tuxfamily.org>.
.SH COPYRIGHT
Copyright \(co 2006 Michael Margraf <michael.margraf@alumni.tu-berlin.de>
.PP
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH AUTHORS
Written by Toyoyuki Ishikawa <ishikawa@conceptkeeper.net> and Michael
Margraf <michael.margraf@alumni.tu-berlin.de>.


ChangeLog
2010-03-25 Renamed project to Caneda
	* Ported all clases and sources to new project name.

2007-10-10 Bastien Roucaries initiated api documentation.
	* qucs/node.h, qucs/node.cpp, qucs/wire.h, qucs/wire.cpp,
	qucs/wireline.h, qucs/item.h, qucs/item.cpp: Add doxygen comments.

2007-03-05  Martín Ribelotta and Lisandro Damián Nicanor Pérez Meyer
	* Ported qucs-edit to qt4
	* Ported qucs-lib to qt4

2006-09-11  Stefan Jahn  <stefan@lkcc.org>

	* configure.ac:  Don't link using "-s" on MacOSX since it drops lazy
	dynamic symbol information.  Also checking for {moc|uic|lrelease|
	lupdate}-qt4 programs.  Added some more paths for Qt headers and
	libraries.

2006-08-08  Stefan Jahn  <stefan@lkcc.org>

	* qucs-attenuator/qucsattenuator.cpp (slotSetText_Zin):
	Synchronized Zout and Zin when specifying a bridged-tee.

	* qucs-attenuator/attenuatorfunc.cpp (Calc): Fixed computation
	routine for bridged-tee topology.

2006-05-29  Stefan Jahn  <stefan@lkcc.org>

	* configure.ac: Released version 0.0.9 and bumped up to version
	0.0.10.

2006-05-14  Stefan Jahn  <stefan@lkcc.org>

	* configure.ac:  Added -s (strip) to linker flags when no
	debugging enabled.
	Put MinGW32 checks in place, thus it is now possible to cross-
	compile the qucs package using a GNU/Linux box.  Had to separate
	LDFLAGS and LIBS because linker command line order matters (-lqt-mt
	-lqtmain *must* be at the end of the command line), otherwise
	linking fails.  Changed rules in appropriate Makefile.am's.
	Additionally checking for the 'ar' tool.

2006-02-16  Stefan Jahn  <stefan@lkcc.org>

	* qucs-transcalc/main.cpp, qucs-lib/main.cpp,
	qucs-help/main.cpp, qucs-filter/main.cpp, qucs-edit/main.cpp
	(loadSettings): Try loading items from main qucsrc even if
	application specific resource file does not exist.

2006-02-10  Stefan Jahn  <stefan@lkcc.org>

	* configure.ac (paths): Added $QTDIR/lib64 to the library path
	to look Qt libraries for.  For 64-bit GNU/Linux's.

2006-01-25  Stefan Jahn  <stefan@lkcc.org>

	* qucs-filter/qf_cauer.cpp: Fixed band stop filters.  Now
	correctly working and exported into the clipboard.

2006-01-21  Stefan Jahn  <stefan@lkcc.org>

	* configure.ac: Released version 0.0.8 and bumped up to version
	0.0.9.

2006-01-18  Stefan Jahn  <stefan@lkcc.org>

	* qucs-filter/qf_filter.cpp: Lowpass/Highpass/Bandpass of
	Cauer filter now quite like working.

	* qucs-filter/qf_cauer.cpp: Fixed bug in bandpass/bandstop
	generation.

2006-01-16  Stefan Jahn  <stefan@lkcc.org>

	* qucs-filter/qf_filter.cpp: Preliminary support for Cauer
	filters.  Many bugs here...

2006-01-14  Stefan Jahn  <stefan@lkcc.org>

	* configure.ac: Checking for Qt programs also in the
	$QTDIR/bin path.

2006-01-07  Stefan Jahn  <stefan@lkcc.org>

	* qucs-filter/qucsfilter.cpp (calculateFilter): Built in
	the cauer filter code by Vincent Habchi.

2005-11-15  Michael Margraf <michael.margraf@alumni.tu-berlin.de>

	* show cursor position in qucs-edit
	* extend Bessel coefficient table to order 19

2005-11-08  Michael Margraf <michael.margraf@alumni.tu-berlin.de>

	* fixed small bug in library search

2005-10-18  Stefan Jahn  <stefan@lkcc.org>

	* configure.ac: Fixed overwriting CXXFLAGS and CFLAGS
	environment variables via 'configure' command line.

2005-10-13  Michael Margraf <michael.margraf@alumni.tu-berlin.de>

	* add new libraries: Bridges, LEDs

2005-08-19  Michael Margraf <michael.margraf@alumni.tu-berlin.de>

	* add new components models to library, thanks to
	Leandro D'Archivio

2005-07-23  Stefan Jahn  <stefan@lkcc.org>

	* qucs-lib/qucslib.1: Added manpage for the QucsLib library
	manager program.

	* configure.ac: Finally fixed freebsd[456]* detection.

2005-07-22  Stefan Jahn  <stefan@lkcc.org>

	* configure.ac: Released version 0.0.7 and bumped up to version
	0.0.8.

2005-05-25  Stefan Jahn  <stefan@lkcc.org>

	* configure.ac: Released version 0.0.6 and bumped up to version
	0.0.7.

2005-05-11  Stefan Jahn  <stefan@lkcc.org>

	* qucs*.1: Updated manpages.

2005-03-01  Stefan Jahn  <stefan@lkcc.org>

	* configure.ac: Released version 0.0.5 and bumped up to version
	0.0.6.

2005-02-27  Stefan Jahn  <stefan@lkcc.org>

	* qucs-help/main.cpp, qucs-edit/main.cpp (main): Made the
	QucsHelp and QucsEdit applications relocatable using the
	environment variable QUCSDIR.

2005-02-23  Stefan Jahn  <stefan@lkcc.org>

	* THANKS: Corrected Toyoyuki Ishikawa's name at all locations
	in the project.

2005-02-21  Stefan Jahn  <stefan@lkcc.org>

	* configure.ac: LDFLAGS additions for FreeBSD users.

2005-02-18  Raimund 'Raimi' Jacob  <raimi@lkcc.org>

	* configure.ac: Add another paths to look for Qt header/library in.
	Required for FreeBSD. Use gmake there.

2004-11-24  Raimund 'Raimi' Jacob  <raimi@lkcc.org>

	* autogen.sh: Make aclocal and automake program be overridable
	with environment variable. Need this so i can easily build on
	debian woody, I hope nobody takes offense.
	PS: While Stefan is away, I will commit his stuff.

2004-11-23  Michael Margraf <michael.margraf@alumni.tu-berlin.de>

	* added icon to QucsEdit and QucsHelp

2004-11-11  Stefan Jahn  <stefan@lkcc.org>

	* qucs-edit/qucsedit.cpp (slotAbout): Added proper copyright
	notice to the 'About' dialog.

2004-10-23  Stefan Jahn  <stefan@lkcc.org>

	* configure.ac: Released version 0.0.4 and bumped up to version
	0.0.5.

2004-09-30  Stefan Jahn  <stefan@lkcc.org>

	* qucs-edit/qucsedit.1: Added simple man page for qucsedit.

	* qucs-help/qucshelp.1: Added simple man page for qucshelp.

	* qucs-help/main.cpp (main): Trying to obtain the correct
	language help directory by splitting e.g. 'es_MX' to use the
	'es' directory.

2004-09-15  Stefan Jahn  <stefan@lkcc.org>

	* qucs-help/main.cpp (main): Using locale() string to check
	whether the documentation directory in the specific language
	is available.

	* configure.ac: Made the DOCDIR variable for choosing a
	specific language.

2004-09-04  Stefan Jahn  <stefan@lkcc.org>


caneda-attenuator/canedaattenuator.cpp
/***************************************************************************
                       Caneda Attenuator Synthesis
                           canedaattenuator.cpp
                               ------------
    begin                : Jun 14 2006

 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/

#include "caneda-tools/global.h"
#include "attenuatorfunc.h"
#include "canedaattenuator.h"
#include "helpdialog.h"
#include <QtCore/QString>

#include <QtGui/QPixmap>
#include <QtGui/QGridLayout>
#include <QtGui/QVBoxLayout>
#include <QtGui/QGroupBox>
#include <QtGui/QMenuBar>
#include <QtGui/QMenu>
#include <QtGui/QMessageBox>
#include <QtGui/QLabel>
#include <QtGui/QPushButton>
#include <QtGui/QLineEdit>
#include <QtGui/QComboBox>
#include <QtGui/QValidator>
#include <QtGui/QClipboard>
#include <QtGui/QApplication>
#include <QtGui/QImage>
#include <QtGui/QStatusBar>
#include <QtGui/QFrame>
#include <QtCore/QTimer>

CanedaAttenuator::CanedaAttenuator()
{
   setWindowIcon(QPixmap(Caneda::bitmapDirectory() + "big.caneda.xpm"));
   setWindowTitle(tr("Caneda Attenuator") + " " + Caneda::version);

  QMenuBar *bar = new QMenuBar(this);
  QMenu *fileMenu = bar->addMenu(tr("&File"));
  bar->addSeparator ();
  QMenu *helpMenu = bar->addMenu(tr("&Help"));
  bar->setFixedHeight(bar->sizeHint().height());
  
  fileMenu->addAction(tr("E&xit"), this, SLOT(close()), Qt::CTRL+Qt::Key_Q);
  
  helpMenu->addAction(tr("Help..."), this, SLOT(slotHelpIntro()), Qt::Key_F1);
  helpMenu->addSeparator();
  helpMenu->addAction(tr("&About CanedaAttenuator..."), this, SLOT(slotHelpAbout()), 0);
  helpMenu->addAction(tr("About Qt..."), this, SLOT(slotHelpAboutQt()), 0);
  
  statusBar = new QStatusBar(this);
  statusBar->setFixedHeight(statusBar->sizeHint().height());
  // First set up layouts and button groups
  QVBoxLayout *topLayout = new QVBoxLayout(this);
  topLayout->setMargin(0);
  topLayout->addWidget(bar);
  QFrame *f = new QFrame(this);
  topLayout->addWidget(f);
  //topLayout->addLayout(mainLayout);
  topLayout->addWidget(statusBar);
  QHBoxLayout * mainLayout = new QHBoxLayout(f);
  QGroupBox * TopoGroup = new QGroupBox (tr("Topology"));
  TopoGroup->setAlignment(Qt::AlignHCenter);
  
  mainLayout->addWidget(TopoGroup);
  QVBoxLayout *topoLayout = new QVBoxLayout(TopoGroup);
  ComboTopology = new QComboBox(TopoGroup);//Topology Combobox
  topoLayout->addWidget(ComboTopology);
  pixTopology = new QLabel(TopoGroup);//Pixmap for Topology
  pixTopology->setMinimumSize(250,175);
  topoLayout->addWidget(pixTopology);

  QVBoxLayout *ioLayout = new QVBoxLayout();
  mainLayout->addLayout(ioLayout);
  
  QGroupBox * InputGroup = new QGroupBox (tr("Input"));
  InputGroup->setAlignment(Qt::AlignHCenter);
  ioLayout->addWidget(InputGroup);
  Calculate = new QPushButton(tr("Calculate and put into Clipboard"));
  
  ioLayout->addWidget(Calculate);
  QGroupBox * OutputGroup = new QGroupBox (tr("Output"));
  OutputGroup->setAlignment(Qt::AlignHCenter);
  ioLayout->addWidget(OutputGroup);
  
  QGridLayout * inpLayout = new QGridLayout(InputGroup);
  QGridLayout * outLayout = new QGridLayout(OutputGroup);

  // Set up the widgets now
  ComboTopology->addItems(QStringList() << tr("Pi") << tr("Tee") << tr("Bridged Tee"));
  connect(ComboTopology, SIGNAL(activated(int)),this,SLOT(slotTopologyChanged()));
    
  pixTopology->setPixmap(QPixmap(Caneda::bitmapDirectory() + "att_pi.png"));

  IntVal = new QIntValidator(this);
  DoubleVal = new QDoubleValidator(this);

  LabelAtten = new QLabel(tr("Attenuation:"));
  lineEdit_Attvalue = new QLineEdit();
  lineEdit_Attvalue->setValidator(DoubleVal);
  QLabel *Label1 = new QLabel(tr("dB"));
  inpLayout->addWidget(LabelAtten,0,0);
  inpLayout->addWidget(lineEdit_Attvalue,0,1);
  inpLayout->addWidget(Label1,0,2);
  
  
  LabelImp1 = new QLabel(tr("Zin:"));
  lineEdit_Zin = new QLineEdit();
  lineEdit_Zin->setValidator(DoubleVal);
  connect(lineEdit_Zin, SIGNAL(textChanged(const QString&)), this,
	  SLOT(slotSetText_Zin(const QString&)) );
  QLabel *Label2 = new QLabel(tr("Ohm"), InputGroup);
  inpLayout->addWidget(LabelImp1,1,0);
  inpLayout->addWidget(lineEdit_Zin,1,1);
  inpLayout->addWidget(Label2,1,2);

  LabelImp2 = new QLabel(tr("Zout:"));
  lineEdit_Zout = new QLineEdit();
  lineEdit_Zout->setValidator(DoubleVal);
  connect(lineEdit_Zout, SIGNAL(textChanged(const QString&)), this,
	  SLOT(slotSetText_Zout(const QString&)) );
  QLabel *Label3 = new QLabel(tr("Ohm"));
  inpLayout->addWidget(LabelImp2,2,0);
  inpLayout->addWidget(lineEdit_Zout,2,1);
  inpLayout->addWidget(Label3,2,2);
  
  connect(Calculate, SIGNAL(clicked()), this, SLOT(slotCalculate()));

  
  LabelR1 = new QLabel(tr("R1:"));
  lineEdit_R1 = new QLineEdit(tr("--"));
  lineEdit_R1->setReadOnly(true);
  QLabel *Label4 = new QLabel(tr("Ohm"));
  outLayout->addWidget(LabelR1, 0,0);
  outLayout->addWidget(lineEdit_R1, 0,1);
  outLayout->addWidget(Label4, 0,2);

  LabelR2 = new QLabel(tr("R2:"));
  lineEdit_R2 = new QLineEdit(tr("--"));
  lineEdit_R2->setReadOnly(true);
  QLabel *Label5 = new QLabel(tr("Ohm"));
  outLayout->addWidget(LabelR2, 1,0);
  outLayout->addWidget(lineEdit_R2, 1,1);
  outLayout->addWidget(Label5, 1,2);

  LabelR3 = new QLabel(tr("R3:"));
  lineEdit_R3 = new QLineEdit(tr("--"));
  lineEdit_R3->setReadOnly(true);
  LabelR3_Ohm = new QLabel(tr("Ohm"));
  outLayout->addWidget(LabelR3, 2,0);
  outLayout->addWidget(lineEdit_R3, 2,1);
  outLayout->addWidget(LabelR3_Ohm, 2,2);
  
  // Make sure this is at last, else may cause seg fault
  readSettings();
}

CanedaAttenuator::~CanedaAttenuator()
{
  delete IntVal;
  delete DoubleVal;
}

void CanedaAttenuator::slotHelpIntro()
{
  HelpDialog *d = new HelpDialog(this);
  d->show();
}

void CanedaAttenuator::slotHelpAboutQt()
{
      QMessageBox::aboutQt(this, tr("About Qt"));
}

void CanedaAttenuator::slotHelpAbout()
{
    QMessageBox::about(this, tr("About..."),
                       QString("CanedaAttenuator Version " PACKAGE_VERSION)+
		       tr("\nAttenuator synthesis program\n")+
		       tr("Copyright (C) 2006 by")+" Toyoyuki Ishikawa"
		       "\n"+
		       tr("Copyright (C) 2006 by")+" Stefan Jahn"
		       "\n"
		       "\nThis is free software; see the source for copying conditions."
		       "\nThere is NO warranty; not even for MERCHANTABILITY or "
		       "\nFITNESS FOR A PARTICULAR PURPOSE.\n\n");
}

void CanedaAttenuator::slotSetText_Zin( const QString &text )
{
  if(ComboTopology->currentIndex() == BRIDGE_TYPE) {
    lineEdit_Zout->blockSignals( TRUE );
    lineEdit_Zout->setText( text );
    lineEdit_Zout->blockSignals( FALSE );
  }
}

void CanedaAttenuator::slotSetText_Zout( const QString &text )
{
  if(ComboTopology->currentIndex() == BRIDGE_TYPE) {
    lineEdit_Zin->blockSignals( TRUE );
    lineEdit_Zin->setText( text );
    lineEdit_Zin->blockSignals( FALSE );
  }
}

void CanedaAttenuator::slotTopologyChanged()
{
  switch(ComboTopology->currentIndex())
    {
    case PI_TYPE:
      pixTopology->setPixmap(QPixmap(Caneda::bitmapDirectory() + "att_pi.png"));
      LabelR2->setText(tr("R2:"));
      LabelR3->show();
      lineEdit_R3->show();
      LabelR3_Ohm->show();
      break;
    case TEE_TYPE:
      pixTopology->setPixmap(QPixmap(Caneda::bitmapDirectory() + "att_tee.png"));
      LabelR2->setText(tr("R2:"));
      LabelR3->show();
      lineEdit_R3->show();
      LabelR3_Ohm->show();
      break;
    case BRIDGE_TYPE:
      pixTopology->setPixmap(QPixmap(Caneda::bitmapDirectory() + "att_bridge.png"));
      LabelR2->setText(tr("R4:"));
      LabelR3->hide();
      lineEdit_R3->hide();
      LabelR3_Ohm->hide();
      lineEdit_Zout->setText( lineEdit_Zin->text() );
      break;
    }
}

void CanedaAttenuator::slotCalculate()
{
    CANEDA_Att qatt;
    int result;
    QString * s = NULL;
    struct tagATT Values;


    Values.Topology = ComboTopology->currentIndex();
    Values.Attenuation = lineEdit_Attvalue->text().toDouble();
    Values.Zin = lineEdit_Zin->text().toDouble();
    Values.Zout = lineEdit_Zout->text().toDouble();
    result = qatt.Calc(&Values);

    if(result != -1)
    {
      statusBar->showMessage(tr("Result:")+" "+tr("Success!"));
      lineEdit_R1->setText(QString::number(Values.R1, 'f', 1));
      lineEdit_R2->setText(QString::number(Values.R2, 'f', 1));
      lineEdit_R3->setText(QString::number(Values.R3, 'f', 1));
      
      s = qatt.createSchematic(&Values);
      if(!s) return;
      
      QClipboard *cb = QApplication::clipboard();
      cb->setText(*s);
      delete s;
    }
    else
    {
      statusBar->showMessage(tr("Error: Set Attenuation more than %1 dB").arg(QString::number(Values.MinimumATT, 'f', 3)));
      lineEdit_R1->setText(tr("--"));
      lineEdit_R2->setText(tr("--"));
      lineEdit_R3->setText(tr("--"));
    }

}

void CanedaAttenuator::readSettings()
{
   QSettings settings;

   settings.beginGroup("MainWindow");
   resize(settings.value("size", QSize(400, 400)).toSize());
   move(settings.value("pos", QPoint(200, 200)).toPoint());
   settings.endGroup();

   settings.beginGroup("SavedInputValues");
   lineEdit_Attvalue->setText(settings.value("Attenuation","1").toString());
   lineEdit_Zin->setText(settings.value("Zin","50").toString());
   lineEdit_Zout->setText(settings.value("Zout","50").toString());
   ComboTopology->setCurrentIndex(settings.value("Topology",PI_TYPE).toInt());
   slotTopologyChanged();
   settings.endGroup();
}

void CanedaAttenuator::writeSettings()
{
   QSettings settings;

   settings.beginGroup("MainWindow");
   settings.setValue("size", size());
   settings.setValue("pos", pos());
   settings.endGroup();

   settings.beginGroup("SavedInputValues");
   settings.setValue("Attenuation",lineEdit_Attvalue->text());
   settings.setValue("Zin",lineEdit_Zin->text());
   settings.setValue("Zout",lineEdit_Zout->text());
   settings.setValue("Topology",ComboTopology->currentIndex());
   settings.endGroup();
}

void CanedaAttenuator::closeEvent(QCloseEvent *e)
{
   writeSettings();
   QWidget::closeEvent(e);
}
